src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/PoolWriter.java

```diff
@@ -124,16 +164,100 @@ int putClass(Type t) {
      * or an array type.
      */
     int putClass(ConstantPoolQType t) {
+        // still need to check here if linkage is necessary
         return pool.writeIfNeeded(t);
     }
 
+    Set<Object> referencesBeingLinked = new HashSet<>();
+
     /**
      * Puts a member reference into the constant pool. Valid members are either field or method symbols.
      */
     int putMember(Symbol s) {
+        Attribute.Compound parametricAnno = s.attribute(syms.parametricType.tsym);
+        if (supportParametricVM &&
+                parametricAnno != null &&
+                enclMethodHasLinkageAnno() &&
+                !referencesBeingLinked.contains(s)) {
+            int linkageIndex = generateLinkage(s, parametricAnno);
+            if (linkageIndex > 0) {
+                return linkageIndex;
+            }
+        }
         return pool.writeIfNeeded(s);
     }
 
+    // where
+        boolean enclMethodHasLinkageAnno() {
+            return currentMethSymbol != null && (currentMethSymbol.attribute(syms.linkageMethodType.tsym) != null ||
+                    currentMethSymbol.attribute(syms.linkageClassType.tsym) != null);
+        }
+
+        /* this method will try to generate a Linkage that will eventually be written into the constant pool
+         * if it can't find all the needed information it will bail out, and return -1, to inform the caller
+         * to generate the corresponding legacy entry. If successful it will return the index of the corresponding
+         * Linkage_info constant in the constant pool
+         */
+        int generateLinkage(Object s, Attribute.Compound parametricAnno) {
```

Comment:
This can probably be simplified, by having this method return the constant to be added - e.g.
```
Pool.Constant wrapWithLinkageIfNeeded(...)
```
And then let clients call `writeIfNeeded`, so that if flow is also more consistent with other pool methods.